cmd/erasure-readfile.go

```diff
@@ -22,10 +22,79 @@ import (
 	"github.com/minio/minio/pkg/errors"
 )
 
-// ReadFile reads as much data as requested from the file under the given volume and path and writes the data to the provided writer.
-// The algorithm and the keys/checksums are used to verify the integrity of the given file. ReadFile will read data from the given offset
-// up to the given length. If parts of the file are corrupted ReadFile tries to reconstruct the data.
-func (s ErasureStorage) ReadFile(writer io.Writer, volume, path string, offset, length int64, totalLength int64, checksums [][]byte, algorithm BitrotAlgorithm, blocksize int64) (f ErasureFileInfo, err error) {
+type errIdx struct {
+	idx int
+	err error
+}
+
+func (s ErasureStorage) readConcurrent(volume, path string, offset, length int64,
+	verifiers []*BitrotVerifier) (buffers [][]byte, err error) {
+
+	errChan := make(chan errIdx)
+	stageBuffers := make([][]byte, len(s.disks))
+	buffers = make([][]byte, len(s.disks))
+
+	drainErrors := func(count int) {
+		for i := 0; i < count; i++ {
+			_ = <-errChan
+		}
+	}
+	readDisk := func(i int) {
+		stageBuffers[i] = make([]byte, length)
+		disk := s.disks[i]
+		if disk == OfflineDisk {
+			errChan <- errIdx{i, errors.Trace(errDiskNotFound)}
+			return
+		}
+		_, rerr := disk.ReadFile(volume, path, offset, stageBuffers[i], verifiers[i])
+		errChan <- errIdx{i, rerr}
+	}
+
+	var finishedCount, successCount, failedCount, launchIndex int
+	requiredBlocks := s.dataBlocks
+
+	for successCount < s.dataBlocks && failedCount <= s.parityBlocks {
+		for ; launchIndex < requiredBlocks; launchIndex++ {
+			go readDisk(launchIndex)
+		}
+		select {
+		case errVal := <-errChan:
+			finishedCount++
+			if errVal.err != nil {
+				// TODO: meaningfully log the disk read error
+
+				failedCount++
+				// A disk failed to return data, so we
+				// request an additional disk if possible
+				if requiredBlocks < s.dataBlocks+s.parityBlocks {
+					requiredBlocks++
+				}
+			} else {
+				successCount++
+				buffers[errVal.idx] = stageBuffers[errVal.idx]
+				stageBuffers[errVal.idx] = nil
+			}
+		}
+	}
+	if successCount == s.dataBlocks {
+		go drainErrors(launchIndex - finishedCount)
```

Comment:
Isn't launchIndex always equal to finishedCount ?

Reply:
Good point! I am further simplifying this.